File: LeetCode/Javascript/Linkedlist/mergetwoSortedList.js
// Merge Two Sorted Lists

        // You are given the heads of two sorted linked lists list1 and list2.

        // Merge the two lists into one sorted list. The list should be made by splicing together the nodes of the first two lists.

        // Return the head of the merged linked list.

// Input: list1 = [1,2,4], list2 = [1,3,4]
// Output: [1,1,2,3,4,4]

// Example 2:
// Input: list1 = [], list2 = []
// Output: []

// Example 3:
// Input: list1 = [], list2 = [0]
// Output: [0]

function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
 }

var mergeTwoLists = function(list1, list2) {

    // Create a dummy node. It helps us easily build the merged list.
    let dummy = new ListNode(0);

    // 'current' always points to the last node in the merged list.
    let current = dummy;

    // Continue while both lists still have nodes.
    while (list1 !== null && list2 !== null) {

        // Compare the current nodes of both lists.
        if (list1.val <= list2.val) {

            // Attach list1's current node.
            current.next = list1;

            // Move list1 to its next node.
            list1 = list1.next;

        } else {

            // Attach list2's current node.
            current.next = list2;

            // Move list2 to its next node.
            list2 = list2.next;
        }

        // Move current forward because a new node was added.
        current = current.next;
    }

    // One list is finished.
    // Attach whichever list still has remaining nodes.
    if (list1 !== null) {
        current.next = list1;
    } else {
        current.next = list2;
    }

    // Skip the dummy node and return the real head.
    return dummy.next;
};

let list1 = new ListNode(1);
list1.next = new ListNode(2);
list1.next.next = new ListNode(4);

let list2 = new ListNode(1);
list2.next = new ListNode(3);
list2.next.next = new ListNode(4);


let result = mergeTwoLists(list1, list2);

console.log(result)


//O(n)
//space O(1)

File: LeetCode/Go/addTwoNumber/mergetwoSortedList.go
package main

import "fmt"

type SingleListNode struct {
	Val  int
	Next *SingleListNode
}

func MergeTwoLists(list1 *SingleListNode, list2 *SingleListNode) *SingleListNode {
	// Create a dummy node. It helps us easily build the merged list.

	dummy := &SingleListNode{}

	// 'current' always points to the last node in the merged list.

	current := dummy

	// Continue while both lists still have nodes.

	for list1 != nil && list2 != nil {
		// Compare the current nodes of both lists.
		if list1.Val <= list2.Val {
			// Attach list1's current node.
			current.Next = list1

			// Move list1 to its next node.
			list1 = list1.Next
		} else {
			current.Next = list2
			list2 = list2.Next
		}
		// Move current forward because a new node was added.
		current = current.Next
	}

	// One list is finished.
	// Attach whichever list still has remaining nodes.
	if list1 != nil {
		current.Next = list1
	} else {
		current.Next = list2
	}

	// Skip the dummy node and return the real head.
	return dummy.Next

}

func printMergeList() {
	// Create list1: 1 -> 2 -> 4
	list1 := &SingleListNode{Val: 1}
	list1.Next = &SingleListNode{Val: 2}
	list1.Next.Next = &SingleListNode{Val: 4}

	// Create list2: 1 -> 3 -> 4
	list2 := &SingleListNode{Val: 1}
	list2.Next = &SingleListNode{Val: 3}
	list2.Next.Next = &SingleListNode{Val: 4}

	// Merge the lists
	result := MergeTwoLists(list1, list2)

	// Print the merged list
	for result != nil {
		fmt.Printf("%d ", result.Val)
		result = result.Next
	}
}
